drivers/sensor/lis2dw12/lis2dw12.c

```diff
@@ -0,0 +1,366 @@
+}
+
+static inline void lis2dw12_channel_get_acc(struct device *dev,
+					     enum sensor_channel chan,
+					     struct sensor_value *val)
+{
+	int i;
+	u8_t ofs_start, ofs_stop;
+	struct lis2dw12_data *lis2dw12 = dev->driver_data;
+	struct sensor_value *pval = val;
+
+	switch (chan) {
+	case SENSOR_CHAN_ACCEL_X:
+		ofs_start = ofs_stop = 0;
+		break;
+	case SENSOR_CHAN_ACCEL_Y:
+		ofs_start = ofs_stop = 1;
+		break;
+	case SENSOR_CHAN_ACCEL_Z:
+		ofs_start = ofs_stop = 2;
+		break;
+	default:
+		ofs_start = 0; ofs_stop = 2;
+		break;
+	}
+
+	for (i = ofs_start; i <= ofs_stop ; i++) {
+		lis2dw12_convert(pval++, lis2dw12->acc[i], lis2dw12->gain);
+	}
+}
+
+static int lis2dw12_channel_get(struct device *dev,
+				 enum sensor_channel chan,
+				 struct sensor_value *val)
+{
+	switch (chan) {
+	case SENSOR_CHAN_ACCEL_X:
+	case SENSOR_CHAN_ACCEL_Y:
+	case SENSOR_CHAN_ACCEL_Z:
+	case SENSOR_CHAN_ACCEL_XYZ:
+		lis2dw12_channel_get_acc(dev, chan, val);
+		return 0;
+	default:
+		SYS_LOG_DBG("Channel not supported");
+		break;
+	}
+
+	return -ENOTSUP;
+}
+
+static int lis2dw12_config(struct device *dev, enum sensor_channel chan,
+			    enum sensor_attribute attr,
+			    const struct sensor_value *val)
+{
+	switch (attr) {
+	case SENSOR_ATTR_FULL_SCALE:
+		return lis2dw12_set_range(dev, sensor_ms2_to_g(val));
+	case SENSOR_ATTR_SAMPLING_FREQUENCY:
+		return lis2dw12_set_odr(dev, val->val1);
+	default:
+		SYS_LOG_DBG("Acc attribute not supported");
+		break;
+	}
+
+	return -ENOTSUP;
+}
+
+static int lis2dw12_attr_set(struct device *dev, enum sensor_channel chan,
+			      enum sensor_attribute attr,
+			      const struct sensor_value *val)
+{
+	switch (chan) {
+	case SENSOR_CHAN_ACCEL_X:
+	case SENSOR_CHAN_ACCEL_Y:
+	case SENSOR_CHAN_ACCEL_Z:
+	case SENSOR_CHAN_ACCEL_XYZ:
+		return lis2dw12_config(dev, chan, attr, val);
+	default:
+		SYS_LOG_DBG("Attr not supported on %d channel", chan);
+		break;
+	}
+
+	return -ENOTSUP;
+}
+
+static int lis2dw12_sample_fetch(struct device *dev, enum sensor_channel chan)
+{
+	struct lis2dw12_data *lis2dw12 = dev->driver_data;
+	const struct lis2dw12_device_config *cfg = dev->config->config_info;
+	u8_t shift;
+	union {
+		u8_t raw[6];
+		struct {
+			s16_t a_axis[3];
+		};
+	} buf __aligned(2);
+	u8_t tmp;
+
+	if (lis2dw12->hw_tf->read_reg(lis2dw12, LIS2DW12_STATUS_REG, &tmp)) {
+		return -EIO;
+	}
+
+	if (!(tmp & LIS2DW12_STS_XLDA_UP)) {
+		return -EAGAIN;
+	}
+
+	/* fetch raw data sample */
+	if (lis2dw12->hw_tf->read_data(lis2dw12, LIS2DW12_OUT_X_L_ADDR,
+				buf.raw, sizeof(buf)) < 0) {
+		SYS_LOG_DBG("Failed to fetch raw data sample");
+		return -EIO;
+	}
+
+	/* adjust to resolution */
+	if (cfg->pm == LIS2DW12_LOW_POWER_M1) {
+		shift = LIS2DW12_SHIFT_PM1;
+	} else {
+		shift = LIS2DW12_SHIFT_PMOTHER;
+	}
+
+	lis2dw12->acc[0] = sys_le16_to_cpu(buf.a_axis[0] >> shift);
+	lis2dw12->acc[1] = sys_le16_to_cpu(buf.a_axis[1] >> shift);
+	lis2dw12->acc[2] = sys_le16_to_cpu(buf.a_axis[2] >> shift);
```

Comment:
I think you need to do the shift after the byte swap?

Reply:
Please check this

Reply:
This is a piece of code which I didn't pay attention to much. It was written by Mario and I
re-used it as it is (and it worked fine). 
But let me give a closer look to it and your observations.

Reply:
OK, let me add some comments (even for myself):
1. The reason why is working on my h/w setup is because the cpu is Little Endian, so the
    swap macro does not actually swap.
2. I think you are right. If the driver is running on a Big Endian platform the driver has to transform
    data into cpu order, and only then perform the shift. 
    So I'll correct this, thx!!!
3. I just realize that some of ST driver does not have this swap inside, so the won't work correctly
    on Big Endian platforms. I will submit a proper fix. Thx!